source/_posts/angular-signals.md

```diff
@@ -0,0 +1,409 @@
+
+## The concept
+
+The main concept of signals in Angular is the ability to **granularly update parts of the component**. We don't want to render the whole component or its tree, but the exact small part of the component that has been changed. And this is why signals were introduced.
+
+### What is the Signal?
+
+A signal is not something new in Angular. The concept already exists in SolidJS, Preact, or Vue.js. The inspiration was taken from SolidJS creator Ryan Carniato.
+
+The signal is a **trackable value**. If the value of the signal changes, all related dependencies are automatically updated.
+
+```ts
+const counter = signal(0);
+```
+
+It is an object consisting of a getter, setters, and a value that changes over time. This "reactive" value can notify all interested consumers.
+### No side effects
+
+A signal is an argument-free function `() => T` that does not create side effects when we read its value. Why doesn't it create side effects? 
+
+First, let's quickly understand what a side effect is. Side effect occurs when a function uses or relies on code from outside. This means that if you use external code in your function, you create a side effect. The example below is an impure function with one side effect.
+
+```ts
+let counter = 0;
+
+function increment() {
+    return counter++;
+}
+```
+
+* Impure function - function with one or more side effects.
+* Pure function - function without side effects.
+
+Perhaps more than once you have found yourself creating side effects using RxJS streams. I have to admit that it has happened to me e.g. in `tap` or `map` operators. Therefore, it is worth emphasizing that signal is not a stream, but only a pure function. It does not create side effects, and its value is always available synchronously. However, it can lazily recalculate intermediate values.
+
+### How to change the Signal value?
+
+Let's say we want to change our value in a signal. We do this via a setter interface `SettableSignal<T>`, which has 3 methods.
+
+```ts
+set(value: T): void;
+update(updateFn: (value: T) => T): void;
+mutate(mutatorFn: (value: T) => void): void;
+```
+
+1. `set()` will completely replace our value.
+
+    ```ts
+    const counter = signal(0);
+    counter.set(1);
+    ```
+
+2. `update()` will update the value of the signal based on its current (old) value.
+
+    ```ts
+    const counter = signal(0);
+    counter.update(value => value + 1);
+    ```
+
+3. `mutate()` will mutate the current value internally and it is a sugar for `update()`. If we didn't want to do this: `users.update(users => [...users, newUser])`, then we can mutate if we want. It's great for objects.
+    ```ts
+    const users = signal<User[])([]);
+    
+    users.mutate(user => {
+        user.push({name: 'Lucas', age: 33});
+    });
+    ```
```

Comment:
Some weird indentations here. Why do the ticks have the indentation in front of it?

Reply:
I thought it is needed in the list. It seems we dont need it.